### ISSUE TITLE
[BUG Calamares 3.2.60] - Wrong Locale and Timezone logic on the interactive world map

### ISSUE THREAD

**Issue by D3vil0per** [calamares/calamares#2008]
**Describe the bug**
At the opening of Calamares, at the Location pane, each time the user clicks on any point of the map, below in "*The system language will be set to ...*" and "*The numbers and dates locale will be set to. ..*" are set as **English (Antigua & Barbuda)**.

Only if the user clicks on any point of Canada or US or Australia, it takes the correct system language and numbers and dates locale.

**Expected behavior**
"*The system language will be set to ...*" and "*The numbers and dates locale will be set to ...*" should change dynamically according their position on the interactive world map.

**Screenshots and Logs**
This is a screenshot when the user clicks on a random point of the world where the logic does not work:
![image](https://user-images.githubusercontent.com/83867734/177005904-bebeae98-59c8-4f19-8bd9-1953042901b4.png)
You can observe below we still get **English (Antigua & Barbuda)**.

This is a screenshot when the user clicks on the previous only quoted countries where the logic works correctly (US, Canada, Australia):
![image](https://user-images.githubusercontent.com/83867734/177005966-34899375-a10c-410f-b1d2-16f91dff4484.png)
Here it is correct.

The host machine is Arch Linux (5.18.8-arch1-1).

I'm available for further information.

Thanks

**Comment by D3vil0per** [comment]
Note that only the countries with "en_*" locale work. The rest of the world where locale does not start with "en_", are set as the first alphabetically sorted "en_" locale, that is "en_AG" .

**Comment by erikdubois** [comment]
We have this since a week or so. I did not find a solution yet.

**Comment by demmm** [comment]
Reverting this commit:
https://github.com/calamares/calamares/commit/79db04dc2eb00b354044f73c054a94fe2b9f9aae
fixes the issue.
But that leaves wrong values for Valencia

**Comment by demmm** [comment]
The above linked commit changes what is used for language code. Old way using QLocale would set `en` to `en_US`, now it just stays as `en` (bug is the same when selecting Spanish `es`, it now reads just as `es`, QLocale sets that to `es_ES`, so now for `es`, the locale page also sets the first available, which is `es_AR` here).
The way I see it, fix needs to come by adding/adjusting this commit:
https://github.com/calamares/calamares/commit/82b19a631463ff781eb636a819bde47014723cd5

The way that is now set, it tries to match a locale like `en` or `es` with a country code, then looks up all available variants for, say `en' with `nl` as country, which has no exact match, thus using the first available.

I'm no good at c++, so anyone with more knowledge might see quickly how to adjust the above commit.

**Comment by demmm** [comment]
I have something working again, but probably not good enough for a PR:
https://github.com/KaOSx/calamares/commit/a2d3c94fb18a3bc60bbae637cfabd6673abbbedc
so please test as a patch on top of the main (3.3) branch

**Comment by killajoe** [comment]
I will give it a try now..

**Comment by demmm** [comment]
The patch linked, does not fix all yet, like is described in issue 2012.

**Comment by killajoe** [comment]
i can simply confirm what you need to get confirmed ;) 
will need a minute to build it in..

**Comment by killajoe** [comment]
now it is setting numbers and date locales when clicking on the map but system language does not change:
![2022-07-03_20-38](https://user-images.githubusercontent.com/16797647/177052952-ba249a05-616a-4403-8033-81bfc5952719.png)

**Comment by demmm** [comment]
@killajoe 
>now it is setting numbers and date locales when clicking on the map but system language does not change:

That is how how it is supposed to be (and has always been), you select the language at the welcome page, locale page is to adjust timezone and dat & number format.  Would be bad if adjusting timezone would mean a different language then what was selected......

**Comment by killajoe** [comment]
> That is how how it is supposed to be (and has always been), you select the language at the welcome page, locale page is to adjust timezone and date & number format. Would be bad if adjusting timezone would mean a different language then what was selected......

okay but in some "clicks" it doe change the language .. in case when i do select englisch US on first screen and click on NZ (newzealand) it changes to en_NZ   or en_CA when clicking on Canada..

**Comment by demmm** [comment]
Again, that is nothing new, or related to this bug report, `en` adjusting on clicking on a country with a different `en` variant (or `es`, or....) same as always.

**Comment by killajoe** [comment]
so it is working as it should that's fine ;)  i wanted only to report what i got with your patch

**Comment by demmm** [comment]
Config.cpp could work with https://github.com/calamares/calamares/commit/79db04dc2eb00b354044f73c054a94fe2b9f9aae
if the top CMakeLists sets all available languages https://github.com/calamares/calamares/blob/calamares/CMakeLists.txt#L147 to use a four letter code, so it will result in a same output as the previous QLocale call did.
This would mean a lot of changes in Transifex, since all current .ts files will have to get the needed four letter code (so `ca` for example, needs to become `ca_ES`).
Right solution?  If some of you agree, then I will do a PR with this option.

**Comment by killajoe** [comment]
not that I am an expert but seems legit, as this is what we use in all the locale settings anyway.

**Comment by D3vil0per** [comment]
Add also another evidence. Despite I set the system language as en_US and the numbers and dates language as it_IT, during the installation, I get spanish words:
![image](https://user-images.githubusercontent.com/83867734/177598104-f9c48664-1f8f-4080-b717-6a36e4c29be9.png)

**Comment by D3vil0per** [comment]
@demmm just one question, that is not so related to this specific issue. Time ago I disabled the calamares installation logging option but now, when I install the system, and I click on "Toggle log", no log appears anymore. I don't remember where I changed this setting. Do you know which file I should change for enabling again the logging of the installation?

**Comment by killajoe** [comment]
> Right solution? If some of you agree, then I will do a PR with this option.

Will we go for merging this or do we need more testing ?  
I would like to include the fix in a release update. 
I know i can patch it also :)

**Comment by D3vil0per** [comment]
> > Also the system language became Spanish
> > i can check this..
> 
> i do not see this here.. output is Italian if you set on first slide to Italian.. it will not change on settings for keyboard and system

Maybe I set bad during that installation by setting es_US (or something similar) instead of en_US. So it was my fault.

**Comment by adriaandegroot** [comment]
Let's take a step back and write down what we **expect**, what actually **happens** (in detail) in 3.2.59 and 3.2.60, so that we have a base to test against, rather than a bunch of one-off anecdotes.

## Preparation

From an existing checkout of Calamares, let's make two worktrees for the versions of interest: the one called *3.2.60* is really the *3.2-stable* branch, but if we fix things, we'll want it fixed there and then forward-ported to 3.3.

```
 git worktree add 3.2.59 v3.2.59
 git worktree add 3.2.60 3.2.x-stable
```

Now we have some new subdirectories; build them (commands from the top-level checkout of the repository, once again).

```
cmake -S 3.2.59 -B 3.2.59/build && cmake --build 3.2.59/build
cmake -S 3.2.60 -B 3.2.60/build && cmake --build 3.2.60/build
```

In each of those build directories, make a suitable `settings.conf`. I took the default one from the repo, then edited the sequence down to just showing the view modules and not doing an installation. This way I can see the UI and watch GS and check the summary, without actually doing an install or messing around with time-consuming steps like partitioning.

```
sequence:
- show: [ welcome, locale, summary, finished ]
```

Run both of them from the respective `build/` directory with the `-d` flag so they pick up the just-compiled modules. Check the versions of both with the "About Calamares" button (lower-left). Notice, if you have the widget sidebar enabled, that the layout of the buttons is slightly different because the stable-branch one has text alongside the icon (so you can tell them apart without clicking about all the time).

**Comment by adriaandegroot** [comment]
## Trivial Tests

These trivial tests are done on my **FreeBSD** machine. It doesn't have `/etc/locale` or any of those things. There is no `locale.gen`. Calamares is started from a shell where `LANG=C.UTF-8`. I'm going to start Calamares, check the language it comes up in and the language it displays in the *welcome* page, then click Next and check what location is selected and what the system language is indicated. So, in order from startup:

- **expect** English US, **expect** American English in the language-selector
- **do** click *Next*
- **expect** Amsterdam (because there's GeoIP lookup), **expect** American English (`en_US`)

Here's a screenshot of how Calamares comes up (3.2.60 is at the top):

![image](https://user-images.githubusercontent.com/19431457/179733817-0ae8c9c2-6130-40db-b5c4-a76e831e2da5.png)

and after **do** clicking *Next*,

![image](https://user-images.githubusercontent.com/19431457/179734196-6a6a44b3-b37d-44d4-a839-0fac4fa8ab73.png)

Expectations not met. There's an **additional** change: numbers and dates is Australian in 3.2.60, Netherlands (the country that was found) in 3.2.59. So the problem is confirmed, and can be easily demonstrated even on a non-Linux system.

Looking at the logs from 3.2.59: there's not much in there, because the logging of looking-up-locales was only added for 3.2.60. The only thing is this blurb, but it is from the *welcome* module:

```
12:45:28 [6]: void Config::setLocaleIndex(int)
    Index 14 Selected locale "en" 
    .. Loading "calamares-default_en" from "/usr/home/adridg/src/git/calamares/3.2.59/build/src/branding/default/lang" 
    .. Branding using locale: "en" 
```

In 3.2.60, the situation is different because the lookups had logging added to them. There's the same blurb from *welcome*, but also this:

```
12:45:22 [2]: QStringList loadLocales(const QString &)
    WARNING: Cannot open file "/etc/locale.gen" . Assuming the supported languages are already built into the locale archive. 
12:45:22 [6]: QStringList loadLocales(const QString &)
    Read 195 lines 
    ..  After filtering 79 lines 
```

and this

```
12:53:44 [6]: static LocaleConfiguration LocaleConfiguration::fromLanguageAndLocation(const QString &, const QStringList &, const QString &)
    Mapping "en" in "NL" to locale. 
    .. Matching ("en_AU.UTF-8", "en_CA.UTF-8", "en_GB.UTF-8", "en_HK.UTF-8", "en_IE.UTF-8", "en_NZ.UTF-8", "en_PH.UTF-8", "en_SG.UTF-8", "en_US.UTF-8", "en_ZA.UTF-8") 
    .. Exact formats match for language tag "en_AU.UTF-8" 
```

**Comment by adriaandegroot** [comment]
## Getting Better Logging

The 3.2.59 log is a bit disappointing. But, because Calamares has a tradition of tagged-per-module commits, and small self-contained commits, we can pull in a handful of post-3.2.59 commits into that worktree, build it, and get the same kind of logging as 3.2.60. These five will do:

d80661211cb5477932d6ad8e38f42df2c0ce2824 
2478570debbe7b08c0dc4110d2962c238bc44095 
79db04dc2eb00b354044f73c054a94fe2b9f9aae
b7dcf55d4cf39b7611c10aa0946bed918a739abc 
b8b7b10bfd9707702d6de16a77f374488de856af

After doing that, 3.2.59 shows **part** of the same problem as 3.2.60. The system language is now Australian, although numbers-and-dates are still Netherlands. Demmm had mentioned 79db04dc2eb00b354044f73c054a94fe2b9f9aae before as a bit sketchy; the commit message seems innocuous enough, but let's take it out again. The other four commits really are nothing but adding-logging.

Looking at the log without the offending commit, we see:

```
13:45:59 [6]: static LocaleConfiguration LocaleConfiguration::fromLanguageAndLocation(const QString &, const QStringList &, const QString &)
    Mapping "en_US" in "NL" to locale. 
    .. Matching ("en_AU.UTF-8", "en_CA.UTF-8", "en_GB.UTF-8", "en_HK.UTF-8", "en_IE.UTF-8", "en_NZ.UTF-8", "en_PH.UTF-8", "en_SG.UTF-8", "en_US.UTF-8", "en_ZA.UTF-8") 
```

and with that offending commit, it looks like the 3.2.60 version: it's then mapping "en", and the matching code does the thing on that rather than "en_US".

**So** at least one problem is identified: going from `QLocale().name()` to using the GS `LANG` key changes the strings looked-up. But let's take a different scenario.

## Valencia

- **expect** English US, **expect** American English in the language-selector
- **do** pick *Catalan (Valencia)* from the language-selector 
- **expect** the to change to Catalan (in the Catalan translation, the translation of "welcome" has gender variants "benvingut" and "benvinguda", while the Valencian variant uses ony  "benvingut", but phrases the welcome message differently)
- **do** click *Next*
- **expect** Amsterdam (because there's GeoIP lookup), **expect** Catalan (ca@valencia)

In 3.2.59 (wthout offending commit): selects `ca_ES` on FreeBSD (possibly because it doesn't know of `ca@valencia` variants at all). In 3.2.60: selects American English, which isn't useful at **all** and neither are correct.

**So** we have identified one problem with the `QLocale` approach: it doesn't know about variants sufficiently well.

## Serbian

This has two alphabet variants:

- **expect** English US, **expect** American English in the language-selector
- **do** pick *Serbian* (Cyrillic) from the language-selector 
- **expect** the to change to Serbian, in Cyrillic
- **do** click *Next*
- **expect** Amsterdam, **expect** Serbian (sr_RS)

This scenario is ok. But in Latin, where there's a peculiar mismatch between Calamares language names and QLocale:

- **expect** English US, **expect** American English in the language-selector
- **do** pick *Serbian* (Latin) from the language-selector 
- **expect** the to change to Serbian, in Latin
- **do** click *Next*
- **expect** Amsterdam, **expect** Serbian (sr_RS@latn)

We still get the Cyrillic choice rather than the Latin alphabet, so that's a mismatch here.

**Comment by adriaandegroot** [comment]
## Changing Host OS

Re-doing these tests on a KDE neon system -- which has a lot more locales available than FreeBSD does -- yields the following. The neon version is called User Edition 5.25, based on Ubuntu 20.04.

- ** 3.2.59**
  - English: en_US
  - Catalan: ca_ES
  - Valencian: ca_ES (wrong)
  - Serbian: sr_RS
  - Serbian Latin: sr_RS (wrong)
- **3.2.60**
  - English: en_AG (wrong)
  - Catalan: ca_AD (arguable, wrong)
  - Valencian: ca_ES@valencia
  - Serbian: sr_ME (wrong, Latin in Montenegro)
  - Serbian Latin: sr_RS@latin (right, except it is labeled in Cyrillic which is misleading)

One test not mentioned so far is British English (en_GB). This works in both versions.

Now we have a half dozen test cases (2 English, 2 Catalan, 2 Serbian) and can throw in, say, 2 Portuguese and German for good measure. We also know that the available languages are different in KDE neon and in FreeBSD, so tests need to take those two language lists into account. Good tests will check that each test-case maps correctly with both sets of supported locales (insofar as it is possible: FreeBSD has no `ca@valencia` so it will have to map that to `ca_ES`).

**Comment by geckolinux** [comment]
> Expectations not met. There's an **additional** change: numbers and dates is Australian in 3.2.60

@adriaandegroot  This is what I'm also seeing on openSUSE Tumbleweed and Leap 15.4 with Calamares 3.2.60. Most regions in Asia, Africa, and South America set the locale to Australian English. Do you need any additional from me to help debug?

**Comment by geckolinux** [comment]
> Reverting this commit: [79db04d](https://github.com/calamares/calamares/commit/79db04dc2eb00b354044f73c054a94fe2b9f9aae) fixes the issue. But that leaves wrong values for Valencia

@demmm  Thanks for this, so this will revert it to the previous behavior? This is valid for 3.2.60?

**Comment by adriaandegroot** [comment]
For the record, there is also a branch *issue-2008a* where I'm going to try a different approach to matching.

**Comment by adriaandegroot** [comment]
The branches have gone away and something has merged to *calamares*, I'm going to cherry-pick it into a new issue branch for 3.2 for brief testing.

**Comment by adriaandegroot** [comment]
Fixed in *calamares* branch with 7a262362df6a11ad22d6c2cf2251b54ac7064e99

Fixed in 3.2.61 with f277cbb1e29f81cf8822888acd12e664538dac64 (cherry-picks a bunch of *calamares* branch into one).
